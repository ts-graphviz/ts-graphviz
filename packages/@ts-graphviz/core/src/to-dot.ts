import { DotObjectModel } from '@ts-graphviz/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '@ts-graphviz/ast';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Subgraph}, {@link Node}, and {@link Edge}
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
