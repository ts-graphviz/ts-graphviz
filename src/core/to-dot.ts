import { DotObjectModel } from '#lib/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '#lib/ast';

/**
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  convert?: ConvertFromModelOptions;
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Node}, and {@link Edge}
 * @param options
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
