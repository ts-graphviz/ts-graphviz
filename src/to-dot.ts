import { DotObjectModel } from './common/index.js';
import { stringify, fromModel, PrintOptions } from './ast/index.js';
import { ConvertOptions } from './ast/index.js';

export interface ToDotOptions {
  convert?: ConvertOptions;
  print?: PrintOptions;
}

export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
