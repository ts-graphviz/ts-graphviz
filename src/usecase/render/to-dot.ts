import { Renderer, Dot } from './renderer.js';

export function toDot(object: Dot): string {
  const renderer = new Renderer();
  return renderer.render(object);
}
