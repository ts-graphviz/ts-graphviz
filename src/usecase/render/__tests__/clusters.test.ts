import { EdgeTargetTuple, ISubgraph, Subgraph } from '../../../model/index.js';
import { toDot } from '../to-dot.js';

describe('Subgraph rendering', () => {
  let subgraph: ISubgraph;
  beforeEach(() => {
    subgraph = new Subgraph('test');
  });

  describe('subgraph with comment', () => {
    test('single line comment', () => {
      subgraph.comment = 'this is comment.';
      expect(toDot(subgraph)).toMatchSnapshot();
    });

    test('multi line comment', () => {
      subgraph.comment = 'this is comment.\nsecond line.';
      expect(toDot(subgraph)).toMatchSnapshot();
    });

    describe('label attribute behavior', () => {
      it('plain text label to be quoted by double quotation', () => {
        subgraph.attributes.graph.set('label', 'this is test for graph label');
        subgraph.attributes.edge.set('label', 'this is test for edge label');
        subgraph.attributes.node.set('label', 'this is test for node label');
        expect(toDot(subgraph)).toMatchSnapshot();
      });

      it('html like', () => {
        subgraph.attributes.graph.set('label', '<<B>this is test for graph label</B>>');
        subgraph.attributes.edge.set('label', '<<U>this is test for edge label</U>>');
        subgraph.attributes.node.set('label', '<<I>this is test for node label</I>>');
        expect(toDot(subgraph)).toMatchSnapshot();
      });
    });

    describe('addXxx existXxx removeXxx APIs', () => {
      it('Edge operation methods works', () => {
        const nodes = ['node1', 'node2'].map((id) => subgraph.createNode(id)) as EdgeTargetTuple;
        const edge = subgraph.createEdge(nodes);
        expect(subgraph.existEdge(edge)).toBe(true);
        expect(toDot(subgraph)).toMatchSnapshot();
      });

      it('Subgraph operation methods works', () => {
        const sub = subgraph.createSubgraph('sub');
        expect(subgraph.existSubgraph(sub)).toBe(true);
        expect(toDot(subgraph)).toMatchSnapshot();
      });
    });
  });
});
