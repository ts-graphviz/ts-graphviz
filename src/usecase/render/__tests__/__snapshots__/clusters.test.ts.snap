// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subgraph rendering subgraph with comment addXxx existXxx removeXxx APIs Edge operation methods works 1`] = `
"subgraph \\"test\\" {
  \\"node1\\";
  \\"node2\\";
  \\"node1\\" -- \\"node2\\";
}"
`;

exports[`Subgraph rendering subgraph with comment addXxx existXxx removeXxx APIs Subgraph operation methods works 1`] = `
"subgraph \\"test\\" {
  subgraph \\"sub\\" {
  }
}"
`;

exports[`Subgraph rendering subgraph with comment label attribute behavior html like 1`] = `
"subgraph \\"test\\" {
  graph [
    label = <<B>this is test for graph label</B>>,
  ];
  edge [
    label = <<U>this is test for edge label</U>>,
  ];
  node [
    label = <<I>this is test for node label</I>>,
  ];
}"
`;

exports[`Subgraph rendering subgraph with comment label attribute behavior plain text label to be quoted by double quotation 1`] = `
"subgraph \\"test\\" {
  graph [
    label = \\"this is test for graph label\\",
  ];
  edge [
    label = \\"this is test for edge label\\",
  ];
  node [
    label = \\"this is test for node label\\",
  ];
}"
`;

exports[`Subgraph rendering subgraph with comment multi line comment 1`] = `
"// this is comment.
// second line.
subgraph \\"test\\" {
}"
`;

exports[`Subgraph rendering subgraph with comment single line comment 1`] = `
"// this is comment.
subgraph \\"test\\" {
}"
`;
